src/core/action-creator.ts

```diff
@@ -161,118 +63,159 @@ export function createActions(flux: FluxCapacitor) {
-
-      receiveProducts: (products: Store.Product[]): Actions.Thunk<Actions.ReceiveProducts> =>
-        thunk(Actions.RECEIVE_PRODUCTS, products, metadata),
-
-      receiveCollectionCount: (collection: string, count: number): Actions.Thunk<Actions.ReceiveCollectionCount> =>
-        thunk(Actions.RECEIVE_COLLECTION_COUNT, { collection, count }, metadata),
-
-      receiveNavigations: (navigations: Store.Navigation[]): Actions.Thunk<Actions.ReceiveNavigations> =>
-        thunk(Actions.RECEIVE_NAVIGATIONS, navigations, metadata),
-
-      receivePage: (page: Actions.Payload.Page): Actions.Thunk<Actions.ReceivePage> =>
-        thunk(Actions.RECEIVE_PAGE, page, metadata),
-
-      receiveTemplate: (template: Store.Template): Actions.Thunk<Actions.ReceiveTemplate> =>
-        thunk(Actions.RECEIVE_TEMPLATE, template, metadata),
-
-      receiveRecordCount: (recordCount: number): Actions.Thunk<Actions.ReceiveRecordCount> =>
-        thunk(Actions.RECEIVE_RECORD_COUNT, recordCount, metadata),
+        }),
+
+      deselectRefinement: (navigationId: string, index: number): Actions.DeselectRefinement =>
+        action(Actions.DESELECT_REFINEMENT, { navigationId, index }, {
+          ...metadata,
+          validator: {
+            payload: {
+              func: (_, state) => Selectors.isRefinementSelected(state, navigationId, index),
+              msg: 'navigation does not exist or refinement is not selected'
+            }
+          }
+        }),
+
+      selectCollection: (id: string): Actions.SelectCollection =>
+        action(Actions.SELECT_COLLECTION, id, {
+          ...metadata,
+          validator: {
+            payload: {
+              func: (_, state) => Selectors.collection(state) !== id,
+              msg: 'collection is already selected'
+            }
+          }
+        }),
+
+      selectSort: (index: number): Actions.SelectSort =>
+        action(Actions.SELECT_SORT, index, {
+          ...metadata,
+          validator: {
+            payload: {
+              func: (_, state) => Selectors.sortIndex(state) !== index,
+              msg: 'sort is already selected'
+            }
+          }
+        }),
+
+      updatePageSize: (size: number): Actions.UpdatePageSize =>
+        action(Actions.UPDATE_PAGE_SIZE, size, {
+          ...metadata,
+          validator: {
+            payload: {
+              func: (_, state) => Selectors.pageSize(state) !== size,
+              msg: 'page size is already selected'
+            }
+          }
+        }),
+
+      updateCurrentPage: (page: number): Actions.UpdateCurrentPage =>
+        action(Actions.UPDATE_CURRENT_PAGE, page, {
+          ...metadata,
+          validator: {
+            payload: {
+              func: (_, state) => page !== null && Selectors.page(state) !== page,
+              msg: 'page size is already selected'
+            }
+          }
+        }),
+
+      updateDetails: (id: string, title: string): Actions.UpdateDetails =>
```

Comment:
my only worry is that if I add an ID validator then it might affect the behaviour of our presets. I will try this in a separate PR